### ISSUE TITLE
Expose the DIRECT algorithm from NLopt

### ISSUE THREAD

**Issue by Tomalwo** [esa/pagmo2#356]
It would the great to expose the DIRECT algorithm that is available in NLopt.

**Comment by darioizzo** [comment]
This is related to #77. I do not remember why DIRECT was not exposed back then when @bluescarni made the nlopt wrappers, probably because pygmo did not have integer capabilities back then? or maybe the version of NLOP we were using did not include it? Or some other, more fundamental reason?

In any case if its possible it should not be too difficult, but we are currently busy with other various requests and improvements, so any help on this one would be appreciated, even if to just tell us that its not possible for some reason ....

**Comment by bluescarni** [comment]
When we first wrote the NLopt wrappers the idea was to focus on the local optimisation capabilities. At the time, NLopt had only a few global optimisation algorithms, but it looks like they added more in recent versions.

As @darioizzo puts it, it should not be too difficult to extend the NLopt wrappers to enable the use of the global optimisation algorithms (I believe there is a bullet point about this in #72). We'll get there eventually, but we don't have an ETA at the moment.

**Comment by darioizzo** [comment]
Thats bad, it would destroy the whole "island / migration" stuff. And also the logic of it all ....

**Comment by bluescarni** [comment]
To be clear, it's not all the global optimisation algorithms that have this behaviour, only some of them. So perhaps the compromise here is to wrap only those who let us use our initial population data.

**Comment by Tomalwo** [comment]
> One possible inconvenience is that the stochastic global optimisation algorithms seemingly force the random generation of the initial population, and thus there seems to be no way to tell NLopt to use pagmo's population data instead.
> 

If I'm not wrong, the DIRECT algorithm is deterministic, so it wouldn't be an issue for that particular algorithm. (It would we for others though, e.g., CRS2.)

**Comment by darioizzo** [comment]
Why would in not be an issue? Even if deterministic it depends on an initial population right? Or its solution depends ony on the problem domain and not on an initial guess?

**Comment by Tomalwo** [comment]
DIRECT doesn't depend on an starting point, if that makes sense.
________________________________

**Comment by darioizzo** [comment]
Ok then indeed its possible and relatively easy to add. It is not an algorithm in the spirit of the island model, nor of migration etc.. So its main use would be on a single thread / process via ```pop = direct.evolve(pop)```. But as long as we explicitly document this I am fine with it.

**Comment by darioizzo** [comment]
Also, if I remember correctly, DIRECT can take quite some time to complete .... so its probably good to check how to stop it or how to log its progress ....

**Comment by Tomalwo** [comment]
You're right, it doesn't make sense with the island model (unfortunately). Nevertheless, I've found it to be quite effective when the number of variables is not too large.
________________________________
